File: typescript/conformance/src/encode_roundtrip.ts
// Copyright 2026 Avala AI
// SPDX-License-Identifier: Apache-2.0

/**
 * Conformance runner: encode.
 *
 * Decode a variant, re-encode the gaussians it yields with the reference preset, and write
 * the result. The gate around this (tests/conformance/encode_roundtrip.py) re-encodes the
 * same variant with the Rust reference and requires the Python decoder to read both files
 * identically. TypeScript is a genuine second encoder, not a binding, so that agreement is a
 * real cross-implementation claim: two encoders quantizing onto the same grids and building
 * the same chunk tree produce files that decode to the same scene.
 *
 * The preset mirrors rust/conformance/src/bin/encode_gaussians.rs. Byte layout — deflate,
 * gaussian order within a chunk — legitimately differs between two encoders, so the gate
 * compares decoded content rather than bytes.
 *
 * Usage: encode_roundtrip.js <in.4dgs> <out.4dgs> [sh-bit-depths]
 */

import { unlinkSync, writeFileSync } from "node:fs";
import { spawnSync } from "node:child_process";
import { isDeepStrictEqual } from "node:util";

import { decodeScene, encodeScene, type GaussianInput } from "@4dgs/core";
import { FileHandleReadable } from "@4dgs/nodejs";

function parseDepths(spec: string): number[] {
  return spec.split(",").map((part) => {
    const value = Number.parseInt(part.trim(), 10);
    if (!Number.isInteger(value) || value < 0 || value > 255) {
      throw new Error(`${spec}: not a comma-separated list of bit depths`);
    }
    return value;
  });
}

function pythonSummary(path: string): Record<string, unknown> {
  const result = spawnSync("python3", ["python/conformance/decode_streamed.py", path], {
    encoding: "utf8",
    maxBuffer: 1024 * 1024,
  });
  if (result.status !== 0) {
    throw new Error(`Python refused the TypeScript objects proof for ${path}:\n${result.stderr}`);
  }
  return JSON.parse(result.stdout) as Record<string, unknown>;
}

const membershipScript = String.raw`
import hashlib, json, os, sys
root = os.getcwd()
sys.path.insert(0, os.path.join(root, "python", "fourdgs"))
sys.path.insert(0, os.path.join(root, "tests", "conformance"))
import fourdgs
from canonical import _stable_order

claims = []
for path in sys.argv[1:]:
    scene = fourdgs.read(path)
    ids = scene.gaussians.object_id
    if ids is None:
        claims.append({"count": 0, "sha256": None})
        continue
    digest = hashlib.sha256()
    block = bytearray()
    for index in _stable_order(scene.gaussians):
        block += int(ids[index]).to_bytes(4, "little", signed=False)
        if len(block) >= 64 * 1024:
            digest.update(block)
            block.clear()
    digest.update(block)
    claims.append({"count": len(ids), "sha256": digest.hexdigest()})
print(json.dumps(claims))
`;

function pythonMembership(paths: string[]): unknown[] {
  const result = spawnSync("python3", ["-c", membershipScript, ...paths], {
    encoding: "utf8",
    maxBuffer: 1024 * 1024,
  });
  if (result.status !== 0) {
    throw new Error(`Python refused the complete membership proof:\n${result.stderr}`);
  }
  return JSON.parse(result.stdout) as unknown[];
}

async function run(input: string, output: string, depths: number[]): Promise<void> {
  const source = await FileHandleReadable.open(input);
  let scene;
  try {
    scene = await decodeScene(source, { blockSize: 8 * 1024 });
  } finally {
    await source.close();
  }
  const g = scene.gaussians;

  const gaussians: GaussianInput = {
    count: g.count,
    positions: g.positions,
    scales: g.scales,
    rotations: g.rotations,
    colors: g.colors,
    motions: g.motions,
    muT: g.muT,
    sigmaT: g.sigmaT,
    winLo: g.winLo,
    winHi: g.winHi,
    sh: g.sh ? g.sh.values : null,
    shDegree: g.sh ? g.sh.degree : 0,
    shCoefficients: g.sh ? g.sh.coefficients : 0,
    objectId: g.objectId,
  };

  // The ordinary agreement preset below is deliberately gaussian-only, so an
  // `objects` source must refuse once its Object Table is omitted. Before that
  // negative assertion, positively exercise the complete writer surface: emit
  // the decoded table/tracks, then have the independent Python implementation
  // decode both files and compare the object records (including sampled track
  // poses) and membership. The temporary file is bounded to one encoded scene and
  // removed before this runner returns.
  if (scene.header.profile === "objects") {
    const proof = `${output}.objects-proof.4dgs`;
    try {
      const proofBytes = await encodeScene(gaussians, scene.header.durationSec, {
        cutoff: scene.header.cutoff,
        maxDepth: 4,
        minChunkGaussians: 8,
        writeIndex: true,
        writeStatistics: true,
        writeSummaryOffsets: true,
        writeCrc: true,
        shBands: 3,
        shBitDepths: depths,
        profile: scene.header.profile,
        attributes: Object.fromEntries(scene.header.attributes),
        objects: scene.objects,
      });
      writeFileSync(proof, proofBytes);
      const sourceSummary = pythonSummary(input);
      const proofSummary = pythonSummary(proof);
      const sourceSample = sourceSummary.sample as Record<string, unknown>;
      const proofSample = proofSummary.sample as Record<string, unknown>;
      const [sourceMembership, proofMembership] = pythonMembership([input, proof]);
      const claims = {
        objects: proofSummary.objects,
        objectIds: proofSample.objectIds,
        completeMembership: proofMembership,
      };
      const expected = {
        objects: sourceSummary.objects,
        objectIds: sourceSample.objectIds,
        completeMembership: sourceMembership,
      };
      if (!isDeepStrictEqual(claims, expected)) {
        throw new Error(
          `Python decoded a different object layer from the TypeScript proof:\n` +
            `expected ${JSON.stringify(expected)}\nactual ${JSON.stringify(claims)}`,
        );
      }
    } finally {
      try {
        unlinkSync(proof);
      } catch (error) {
        if ((error as NodeJS.ErrnoException).code !== "ENOENT") throw error;
      }
    }
  }

  // The gaussians-only preset, matching the Rust reference: a small chunk threshold so the
  // corpus scenes exercise the tree, the whole summary written, the profile and attributes
  // carried through, the library left at the encoder's default.
  const bytes = await encodeScene(gaussians, scene.header.durationSec, {
    cutoff: scene.header.cutoff,
    maxDepth: 4,
    minChunkGaussians: 8,
    writeIndex: true,
    writeStatistics: true,
    writeSummaryOffsets: true,
    writeCrc: true,
    shBands: 3,
    shBitDepths: depths,
    profile: scene.header.profile,
    attributes: Object.fromEntries(scene.header.attributes),
  });

  writeFileSync(output, bytes);
  process.stdout.write(`${g.count} gaussians, ${bytes.length} bytes\n`);
}

const [input, output, depthSpec] = process.argv.slice(2);
if (input === undefined || output === undefined) {
  process.stderr.write("usage: encode_roundtrip.js <in.4dgs> <out.4dgs> [sh-bit-depths]\n");
  process.exit(2);
}
await run(input, output, depthSpec === undefined ? [] : parseDepths(depthSpec));

File: dart/conformance/bin/encode_roundtrip.dart
// Copyright 2026 Avala AI
// SPDX-License-Identifier: Apache-2.0

/// Conformance runner: encode.
///
/// Decode a variant, re-encode the gaussians it yields with the reference
/// preset, and write the result. The gate around this
/// (`tests/conformance/encode_roundtrip.py`) re-encodes the same variant with
/// the Rust reference and requires the Python decoder to read both files
/// identically. Dart is a genuine second encoder rather than a binding, so that
/// agreement is a real cross-implementation claim: two encoders quantizing onto
/// the same grids produce files that decode to the same scene.
///
/// The preset mirrors `rust/conformance/src/bin/encode_gaussians.rs`. Byte
/// layout — how well deflate did, which order gaussians sit in a chunk —
/// legitimately differs between two encoders, so the gate compares decoded
/// content rather than bytes.
///
/// Usage: encode_roundtrip <in.4dgs> <out.4dgs>
library;

import 'dart:io';

import 'package:fourdgs/fourdgs.dart';

/// The source's profile, unless it is one this preset cannot keep.
///
/// A profile is a promise about the file's contents, and this runner writes the
/// gaussians alone — so `objects`, which promises an Object Table, and
/// `capture`, which promises a finite multi-chunk timeline with Statistics, are
/// promises this generic preset cannot keep for every source. Dropping them is
/// the same choice the preset already makes about audio, cameras and
/// attachments: what cannot be reproduced is not claimed.
String _writableProfile(String profile) =>
    profile == 'objects' || profile == 'capture' ? '' : profile;

void _checkSourceGroups(FourdgsGaussianSet source, FourdgsGaussianSet encoded) {
  if ((source.sourceGroup == null) != (encoded.sourceGroup == null)) {
    throw StateError(
      'the source_group stream changed between present and absent',
    );
  }
  if (source.sourceGroup == null) return;
  if (source.sourceIndex == null || encoded.sourceIndex == null) {
    throw StateError(
      'source_group fidelity needs the source_index pairing lane in this encode gate',
    );
  }
  final expected = <int, int>{};
  for (int i = 0; i < source.count; i++) {
    final key = source.sourceIndex![i];
    if (expected.containsKey(key)) {
      throw StateError(
        'source_index $key is not unique enough to pair source_group',
      );
    }
    expected[key] = source.sourceGroup![i];
  }
  for (int i = 0; i < encoded.count; i++) {
    final key = encoded.sourceIndex![i];
    if (!expected.containsKey(key) ||
        expected.remove(key) != encoded.sourceGroup![i]) {
      throw StateError('source_group changed for source_index $key');
    }
  }
  if (expected.isNotEmpty) {
    throw StateError(
      'the encoded source_group stream omitted ${expected.length} rows',
    );
  }
}

/// The gaussians alone: no audio, no camera, no metadata records, no
/// attachments, no provenance. That is the whole point of this baseline — these
/// authoring surfaces write gaussians, and a gate that compared whole scenes
/// would fail on the extras rather than on the encode.
String run(String input, String output) {
  final scene = readFourdgsBytes(File(input).readAsBytesSync());
  final bytes = writeFourdgsBytes(
    scene.gaussians,
    scene.header.durationSec,
    options: FourdgsWriteOptions(
      cutoff: scene.header.cutoff,
      maxDepth: 4,
      minChunkGaussians: 8,
      writeIndex: true,
      writeStatistics: true,
      writeSummaryOffsets: true,
      writeCrc: true,
      shBands: 3,
      sceneProfile: _writableProfile(scene.header.profile),
      attributes: scene.header.attributes,
    ),
  );

  // Two encodes of one scene must be the same bytes. An encoder that iterates a
  // map, or sorts unstably, passes every value-based check and still produces a
  // file that differs between runs — which is a property nobody notices until a
  // build is expected to reproduce.
  final again = writeFourdgsBytes(
    scene.gaussians,
    scene.header.durationSec,
    options: FourdgsWriteOptions(
      cutoff: scene.header.cutoff,
      maxDepth: 4,
      minChunkGaussians: 8,
      writeIndex: true,
      writeStatistics: true,
      writeSummaryOffsets: true,
      writeCrc: true,
      shBands: 3,
      sceneProfile: _writableProfile(scene.header.profile),
      attributes: scene.header.attributes,
    ),
  );
  if (bytes.length != again.length) {
    throw StateError('two encodes of one scene differ in length');
  }
  for (int i = 0; i < bytes.length; i++) {
    if (bytes[i] != again[i]) {
      throw StateError('two encodes of one scene differ at byte $i');
    }
  }

  final reread = readFourdgsBytes(bytes);
  if (reread.gaussians.count != scene.gaussians.count) {
    throw StateError(
      'the encoder wrote ${reread.gaussians.count} gaussians for '
      '${scene.gaussians.count}',
    );
  }
  _checkSourceGroups(scene.gaussians, reread.gaussians);

  File(output).writeAsBytesSync(bytes);
  return '${reread.gaussians.count} gaussians, '
      '${reread.chunkIndex.length} chunks, ${bytes.length} bytes, deterministic';
}

void main(List<String> args) {
  if (args.length != 2) {
    stderr.writeln('usage: encode_roundtrip <in.4dgs> <out.4dgs>');
    exit(2);
  }
  try {
    stdout.writeln(run(args[0], args[1]));
  } catch (error) {
    stderr.writeln(error);
    exit(1);
  }
}
